### PR TITLE
Adding required member functions of abstract container traits

### PR DESCRIPTION
**Pull request by adam2392**
Towards: #27 

cc: @bharath2438 is this something along the lines of what you were thinking to add required API for a user defining their own container traits?

### CODE REVIEW

File: include/xsparse/util/container_traits.hpp [hunk 1]
```diff
@@ -22,6 +22,18 @@ namespace xsparse::util
 
         template <class Key, class Val>
         using Map = TMap<Key, Val>;
+
```
Comment (adam2392) [inline include/xsparse/util/container_traits.hpp:25]: Just static_assert for the presence of these functions rather than adding an abstract api.

**Comment by adam2392** [comment]
@hameerabbasi this is good to review now as well. I just add a few static asserts based on what we need from the `Vec` and `Map` in the levels.

One thing I noticed though is that `Set` is not used. Is this an issue?

**Comment by bharath2438** [comment]
@adam2392 this looks good to me. I guess `std::set` has a `contains` function in C++20 so we might as well ensure there's an implementation for that as well.

**Comment by adam2392** [comment]
> @adam2392 this looks good to me. I guess `std::set` has a `contains` function in C++20 so we might as well ensure there's an implementation for that as well.

Kay added!

What would we even use Set for? It seems it is not used in the level formats?

**Comment by bharath2438** [comment]
> What would we even use Set for? It seems it is not used in the level formats?

`std::unordered_set` can be used for representing unordered levels

**Comment by hameerabbasi** [comment]
This LGTM overall. I'd just static assert these things in the container traits class itself, and make sure that the signature matches as well, rather than just the function existing. Otherwise looks excellent! Great work!

File: include/xsparse/util/container_traits.hpp [hunk 2]
```diff
@@ -22,7 +100,40 @@ namespace xsparse::util
...
+            std::is_same_v<decltype(std::declval<Vec<double>>()[std::declval<std::size_t>()]), decltype(std::declval<Vec<double>>().at(std::declval<std::size_t>()))>,
+            "Vec must have `operator[]` method with the correct signature.");
+
+        static_assert(
+            std::is_same_v<decltype(std::declval<Map<int, std::size_t>>().find(std::declval<std::size_t>())), typename Map<int, std::size_t>::iterator>,
+            "Map must have `find` method with the correct signature.");
+
+        static_assert(
+            std::is_same_v<decltype(std::declval<Set<std::size_t>>().contains(std::declval<std::size_t>())), bool>,
+            "Set must have `contains` method with the correct signature.");
```
Comment (adam2392) [inline include/xsparse/util/container_traits.hpp:135]: @hameerabbasi is this what you meant by static assert directly in the container_traits class?

Comment (hameerabbasi) [reply include/xsparse/util/container_traits.hpp:135]: Yes. Or maybe in the destructor if that doesn't work.

Comment (adam2392) [reply include/xsparse/util/container_traits.hpp]: Kay this works and I tried it out by substituting `std::list` for std::vector and as expected, a compiler-error is raised

Comment (hameerabbasi) [reply include/xsparse/util/container_traits.hpp]: Cool, let's get CI working and then we'll be good to go.

Comment (adam2392) [reply include/xsparse/util/container_traits.hpp]: Cool @hameerabbasi this is now fixed

**Pull request merged**